### PR TITLE
pthreads4w: fix support for mingw on Windows

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **lib/1.0**

This should fix building (and installing) pthreads4w on mingw on Windows

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/pthreads4w/all/conanfile.py [hunk 1]
```diff
@@ -23,16 +24,24 @@ def configure(self):
         del self.settings.compiler.cppstd
 
     def build_requirements(self):
-        if self.settings.compiler != 'Visual Studio':
-            if "CONAN_BASH_PATH" not in os.environ and tools.os_info.detect_windows_subsystem() != 'msys2':
-                self.build_requires('msys2/20190524')
+        if "CONAN_BASH_PATH" not in os.environ:
+            self.build_requires('msys2/20190524')
```
Comment (ericLemanissier) [inline recipes/pthreads4w/all/conanfile.py]: Why is this change needed? What does this fix/improve?

Comment (madebr) [reply recipes/pthreads4w/all/conanfile.py]: This wouldn't add the msys2 requirements when using mingw_installer.
But indeed, it is a bug that it adds the `msys2` requirement even when using the `Visual Studio` compiler.

Comment (ericLemanissier) [reply recipes/pthreads4w/all/conanfile.py]: Not sure I understand. Do you mean `tools.os_info.detect_windows_subsystem()` returns msys2 when only mingw_installer is build_require'd ?

Comment (madebr) [reply recipes/pthreads4w/all/conanfile.py]: My profile:
```
conan profile show mingw8
Configuration for profile mingw8:

[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=8
compiler.threads=posix
compiler.exception=seh
compiler.libcxx=libstdc++
build_type=Release
[options]
[build_requires]
*: mingw_installer/1.0@conan/stable
[env]
```
Shell output:
```
conan create . pthreads4w/3.0.0@ -pr mingw8
[HOOK - conan-center.py] pre_export(): [DEPRECATED GLOBAL CPPSTD (KB-H001)] OK
[HOOK - conan-center.py] pre_export(): [REFERENCE LOWERCASE (KB-H002)] OK
[HOOK - conan-center.py] pre_export(): WARN: [RECIPE METADATA (KB-H003)] Conanfile doesn't have 'topics' attribute.
[HOOK - conan-center.py] pre_export(): [RECIPE METADATA (KB-H003)] OK
[HOOK - conan-center.py] pre_export(): [HEADER_ONLY, NO COPY SOURCE (KB-H005)] OK
[HOOK - conan-center.py] pre_export(): WARN: [FPIC OPTION (KB-H006)] This recipe does not include an 'fPIC' option. Make sure you are using the right casing
[HOOK - conan-center.py] pre_export(): [FPIC OPTION (KB-H006)] OK
[HOOK - conan-center.py] pre_export(): [VERSION RANGES (KB-H008)] OK
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] Total recipe size: 5.05859375 KB
[HOOK - conan-center.py] pre_export(): [RECIPE FOLDER SIZE (KB-H009)] OK
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] exports: None
[HOOK - conan-center.py] pre_export(): [EXPORT LICENSE (KB-H023)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE FOLDER (KB-H024)] OK
[HOOK - conan-center.py] pre_export(): [META LINES (KB-H025)] OK
[HOOK - conan-center.py] pre_export(): [CONAN CENTER INDEX URL (KB-H027)] OK
[HOOK - conan-center.py] pre_export(): [CMAKE MINIMUM VERSION (KB-H028)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - RUN ENVIRONMENT (KB-H029)] OK
[HOOK - conan-center.py] pre_export(): [CONANDATA.YML FORMAT (KB-H030)] OK
[HOOK - conan-center.py] pre_export(): [TEST PACKAGE - NO IMPORTS() (KB-H034)] OK
[HOOK - conan-center.py] pre_export(): [NO AUTHOR (KB-H037)] OK
Exporting package recipe
pthreads4w/3.0.0 exports: File 'conandata.yml' found. Exporting it...
pthreads4w/3.0.0 exports: Copied 1 '.yml' file: conandata.yml
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] Saving conandata.yml: {'sources': {'3.0.0': {'sha256': 'b81136effb7185c77601fe2e0e6ac19bd996912e4814cebdd3010b0fac9e259b', 'url': 'https://sourceforge.net/projects/pthreads4w/files/pthreads4w-code-v3.0.0.zip/download'}}}
[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] New conandata.yml contents: sources:
  3.0.0:
    sha256: b81136effb7185c77601fe2e0e6ac19bd996912e4814cebdd3010b0fac9e259b
    url: https://sourceforge.net/projects/pthreads4w/files/pthreads4w-code-v3.0.0.zip/download

[HOOK - conan-center.py] post_export(): [CONANDATA.YML REDUCE (KB-H031)] OK
pthreads4w/3.0.0: The stored package has not changed
pthreads4w/3.0.0: Exported revision: 8c7a3e061b7c50c2b73ba310311755b6
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.exception=seh
compiler.libcxx=libstdc++
compiler.threads=posix
compiler.version=8
os=Windows
os_build=Windows
[options]
[build_requires]
*: mingw_installer/1.0@conan/stable
[env]

pthreads4w/3.0.0: Forced build from source
pthreads4w/3.0.0 (test package): Installing package
Requirements
    pthreads4w/3.0.0 from local cache - Cache
Packages
    pthreads4w/3.0.0:db5be4a3874fa488325a1208d501b31dcbc45e69 - Build
Build requirements
    mingw_installer/1.0@conan/stable from 'conan-center' - Cache
Build requirements packages
    mingw_installer/1.0@conan/stable:649c186bbe7e6cd995c27c9e2e8ee09fe6c43e00 - Cache

mingw_installer/1.0@conan/stable: Already installed!
[HOOK - conan-center.py] post_package_info(): [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] OK
[HOOK - conan-center.py] post_package_info(): [CMAKE FILE NOT IN BUILD FOLDERS (KB-H019)] OK
pthreads4w/3.0.0: Applying build-requirement: mingw_installer/1.0@conan/stable
pthreads4w/3.0.0: WARN: Build folder is dirty, removing it: C:\Users\maarten\.conan\data\pthreads4w\3.0.0\_\_\build\db5be4a3874fa488325a1208d501b31dcbc45e69
pthreads4w/3.0.0: Copying sources to build folder
pthreads4w/3.0.0: Building your package in C:\Users\maarten\.conan\data\pthreads4w\3.0.0\_\_\build\db5be4a3874fa488325a1208d501b31dcbc45e69
pthreads4w/3.0.0: Generator txt created conanbuildinfo.txt
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] 'fPIC' option not found
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
pthreads4w/3.0.0: Calling build()
pthreads4w/3.0.0: run_in_windows_bash: "C:\Program Files\Git\bin\bash.exe" --login -c ^"cd \^"/c/users/maarten/.conan/data/pthreads4w/3.0.0/_/_/build/db5be4a3874fa488325a1208d501b31dcbc45e69/source_folder\^" ^&^& PATH=\^"/c/users/maarten/.conan/data/mingw_installer/1.0/conan/stable/package/649c186bbe7e6cd995c27c9e2e8ee09fe6c43e00/bin:$PATH\^" ^&^& autoheader ^"
pthreads4w/3.0.0:
----Running------
> "C:\Program Files\Git\bin\bash.exe" --login -c ^"cd \^"/c/users/maarten/.conan/data/pthreads4w/3.0.0/_/_/build/db5be4a3874fa488325a1208d501b31dcbc45e69/source_folder\^" ^&^& PATH=\^"/c/users/maarten/.conan/data/mingw_installer/1.0/conan/stable/package/649c186bbe7e6cd995c27c9e2e8ee09fe6c43e00/bin:$PATH\^" ^&^& autoheader ^"
-----------------
pthreads4w/3.0.0: /usr/bin/bash: autoheader: command not found
pthreads4w/3.0.0:
pthreads4w/3.0.0: ERROR: Package 'db5be4a3874fa488325a1208d501b31dcbc45e69' build failed
pthreads4w/3.0.0: WARN: Build folder C:\Users\maarten\.conan\data\pthreads4w\3.0.0\_\_\build\db5be4a3874fa488325a1208d501b31dcbc45e69
ERROR: pthreads4w/3.0.0: Error in build() method, line 56
        self.run('autoheader', win_bash=True)
        ConanException: Error 127 while executing autoheader
```

I've added the line
```
        print('tools.os_info.detect_windows_subsystem() is', tools.os_info.detect_windows_subsystem())
        import conans.client.tools.oss
        print('OSInfo.uname() is', conans.client.tools.oss.OSInfo.uname())
```
at the top of the `build_requirements()` function.
Its output is:
```
tools.os_info.detect_windows_subsystem() is msys2
OSInfo.uname() is mingw64_nt-10.0-18363
```

So yes, this function returns `msys2` even when running from plain `cmd`.
I guess this is because of `mingw_installer/1.0@conan/stable`.

Comment (ericLemanissier) [reply recipes/pthreads4w/all/conanfile.py]: The problem is that you actually have an msys2 bash available in PATH (C:\Program Files\Git\bin\bash.exe). This is why tools.os_info.detect_windows_subsystem() is returns msys2, and it is the reason for puting this check in the first place: avoiding to download the conan package of msys2 if it is already available in the environment. If you don't want to use this distribution, you can either remove it from path, or build_require msys2 from your profile. If you want to use msys2 provided by git, you can use pacman to install missing autoheader etc.

Comment (madebr) [reply recipes/pthreads4w/all/conanfile.py]: Nice catch.
I didn't notice the `"C:\Program Files\Git\bin\bash.exe" --login XXX` lines in my logs.
My PATH variable indeed had 2 paths related to git: `C:\Program Files\Git\cmd` and `C:\Program Files\Git\bin`. (Maybe this should be added to some FAQ because this is the first time I know of this).

Anyway, removing the second path from my PATH env variable fixed the problem.
I squashed my patch to one patch fixing only the packaging using mingw.
